crates/sargon/src/wrapped_radix_engine_toolkit/low_level/v2/transaction_manifest_v2/instructions_v2/instructions_v2.rs

```diff
@@ -0,0 +1,501 @@
+
+    fn try_from(
+        value: (&Vec<ScryptoInstructionV2>, NetworkID),
+    ) -> Result<Self, CommonError> {
+        let scrypto = value.0;
+        let network_id = value.1;
+
+        // Verify that the instructions has acceptable depth and are compatible
+        _ = instructions_string_from(scrypto, network_id)?;
+
+        Ok(Self {
+            secret_magic: InstructionsSecretMagicV2::from(
+                ScryptoInstructionsV2(scrypto.to_owned().into()),
+            ),
+            network_id,
+        })
+    }
+}
+
+fn instructions_string_from(
+    scrypto_instructions: &Vec<ScryptoInstructionV2>,
+    network_id: NetworkID,
+) -> Result<String, CommonError> {
+    let scrypto_manifest = ScryptoTransactionManifestV2 {
+        instructions: scrypto_instructions.clone(),
+        blobs: Default::default(),
+        children: Default::default(),
+        object_names: Default::default(),
+    };
+    manifest_v2_string_from(scrypto_manifest, network_id)
+}
+
+impl InstructionsV2 {
+    pub fn instructions_string(&self) -> String {
+        instructions_string_from(self.secret_magic.instructions().as_ref(), self.network_id).expect("Should never fail, because should never have allowed invalid instructions")
+    }
+
+    pub fn new(
+        instructions_string: impl AsRef<str>,
+        network_id: NetworkID,
+    ) -> Result<Self> {
+        let network_definition = network_id.network_definition();
+        let blob_provider = ScryptoMockBlobProvider::new();
+
+        scrypto_compile_manifest(
+            instructions_string.as_ref(),
+            &network_definition,
+            blob_provider,
+        )
+        .map_err(|e| extract_error_from_error(e, network_id))
+        .and_then(|manifest: ScryptoTransactionManifestV2| {
+            Self::try_from((manifest.instructions.as_ref(), network_id))
+        })
+    }
+}
+
+#[cfg(test)]
+impl InstructionsV2 {
+    /// Utility function which uses `InstructionsV2::new(<string>, <network_id>)`
+    /// and SHOULD return `Err` if `depth > Instructions::MAX_SBOR_DEPTH`, which
+    /// we can assert in unit tests.
+    pub(crate) fn test_with_sbor_depth(
+        depth: usize,
+        network_id: NetworkID,
+    ) -> Result<Self> {
+        let nested_value = manifest_value_with_sbor_depth(depth);
+        let dummy_address =
+            ComponentAddress::with_node_id_bytes(&[0xffu8; 29], network_id);
+        let instruction = ScryptoInstructionV2::CallMethod(CallMethod {
+            address: TryInto::<ScryptoDynamicComponentAddress>::try_into(
+                &dummy_address,
+            )
+            .unwrap()
+            .into(),
+            method_name: "dummy".to_owned(),
+            args: nested_value,
+        });
+        instructions_string_from(&vec![instruction], network_id)
+            .and_then(|x: String| Self::new(x, network_id))
+    }
+
+    pub(crate) const MAX_SBOR_DEPTH: usize = MANIFEST_SBOR_V1_MAX_DEPTH - 3;
+}
+
+fn extract_error_from_addr(
```

Comment:
Seems to be similar to the function declared in `instructions.rs`, or I am missing something :)?

Reply:
Right. Extracted and moved to `common_error.rs`